statistics/row_sampler.go

```diff
@@ -250,28 +269,44 @@ func (s *baseCollector) ToProto() *tipb.RowSampleCollector {
 	return collector
 }
 
-func (s *baseCollector) FromProto(pbCollector *tipb.RowSampleCollector) {
+func (s *baseCollector) FromProto(pbCollector *tipb.RowSampleCollector, memTracker *memory.Tracker) {
 	s.Count = pbCollector.Count
 	s.NullCount = pbCollector.NullCounts
 	s.FMSketches = make([]*FMSketch, 0, len(pbCollector.FmSketch))
 	for _, pbSketch := range pbCollector.FmSketch {
 		s.FMSketches = append(s.FMSketches, FMSketchFromProto(pbSketch))
 	}
 	s.TotalSizes = pbCollector.TotalSize
-	s.Samples = make(WeightedRowSampleHeap, 0, len(pbCollector.Samples))
+	sampleNum := len(pbCollector.Samples)
+	s.Samples = make(WeightedRowSampleHeap, 0, sampleNum)
+	// consume mandatory memory at the beginning, if exceeds, fast fail
+	if len(pbCollector.Samples) > 0 {
+		rowLen := len(pbCollector.Samples[0].Row)
+		// 24 is the size of datum array, 8 is the size of reference
+		initMemSize := int64(sampleNum) * (int64(rowLen)*types.EmptyDatumSize + EmptyReservoirSampleItemSize + 8)
```

Comment:
L281 only allocates a slice of `*ReservoirRowSampleItem` and no actual `ReservoirRowSampleItem` struct is allocated. So maybe we only need to calculate the memory of the slice.

Reply:
It's a pre-consume, if exceeds the limit, fast fail.

Reply:
comment is already there: "// consume mandatory memory at the beginning, if exceeds, fast fail"

Reply:
Update with more details.